### PR TITLE
feat(DTFS2-6627): gef UI add read only limitation on routes

### PR DESCRIPTION
**Pull request by AlexBramhill**
# Introduction
We are adding a `read-only` role. We wish to limit routes that this role can access on GEF-UI

# Resolution
The majority of this ticket is adding tests to existing api routes. There are a few additions to the actual route code itself

### CODE REVIEW

File: gef-ui/api-tests/test-helpers/storage/storage.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const redis = require('redis');
+const { generateUserSession } = require('../user-session-generator');
+
+const connect = () => {
```
Comment (AlexBramhill) [inline gef-ui/api-tests/test-helpers/storage/storage.js:4]: There is an issue here that occurs when multiple tests are ran (and very rarely when a single test is run). 

I assume it is an async issue with db writing. Will have a look Monday

File: portal/server/routes/middleware/validateRole/index.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 /**
  *
- * @param {array} requiredRoles  (i.e. ['maker'])
- * @param {array} user (i.e. ['checker'] or ['maker', 'checker'])
+ * @param {array} requiredRoles  (i.e. [MAKER])
+ * @param {array} user (i.e. [CHECKER] or [MAKER, 'checker'])
```
Comment (lmarrai-sw) [inline portal/server/routes/middleware/validateRole/index.js:4]:
```suggestion
 * @param {array} requiredRoles  (i.e. ['maker'])
 * @param {array} user (i.e. ['checker'] or ['maker', 'checker'])
```

File: portal/server/routes/middleware/validateRole/index.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ const userRoleIsValid = (requiredRoles, user) => {
 };
 /**
  *
- * @param {object} opts  (i.e. { role: ['maker'] })
+ * @param {object} opts  (i.e. { role: [MAKER] })
```
Comment (lmarrai-sw) [inline portal/server/routes/middleware/validateRole/index.js:21]:
```suggestion
 * @param {object} opts  (i.e. { role: ['maker'] })
```

File: portal/package.json [hunk 4]
```diff
@@ -27,6 +27,7 @@
     "start": "node server/index.js",
     "test": "jest --coverage --verbose --config=jest.config.js",
     "test-dev": "jest --verbose --config=jest.config.js",
+    "test-file": "jest --config=jest.config.js --verbose --testMatch",
```
Comment (lmarrai-sw) [inline portal/package.json:30]: Should we document this new command in the README?

File: gef-ui/server/routes/unissued-facilities.js [hunk 5]
```diff
@@ -9,31 +9,41 @@ const {
...
+// TODO: DTFS2-6627 - Check whether this needs validation
 // get change issued facility to unissued from application preview
-router.get('/application-details/:dealId/unissued-facilities/:facilityId/change-to-unissued', validateToken, (req, res) =>
-  changeIssuedToUnissuedFacility(req, res));
+router.get(
+  '/application-details/:dealId/unissued-facilities/:facilityId/change-to-unissued',
+  [validateRole({ role: [MAKER] }), validateBank, validateToken],
+  (req, res) => changeIssuedToUnissuedFacility(req, res),
+);
+// TODO: DTFS2-6627 - Check whether this needs validation
```
Comment (lmarrai-sw) [inline gef-ui/server/routes/unissued-facilities.js:41]: Flagging TODO

File: gef-ui/server/routes/unissued-facilities.js [hunk 6]
```diff
@@ -9,31 +9,41 @@ const {
...
 // get change unissued facility from application preview
-router.get('/application-details/:dealId/unissued-facilities/:facilityId/change', [validateToken, validateBank, validateRole({ role: ['maker'] })], (req, res) =>
+router.get('/application-details/:dealId/unissued-facilities/:facilityId/change', [validateToken, validateBank, validateRole({ role: [MAKER] })], (req, res) =>
   changeUnissuedFacilityPreview(req, res));
 // post change unissued facilities to issued from application preview
-router.post('/application-details/:dealId/unissued-facilities/:facilityId/change', [validateToken, validateBank, validateRole({ role: ['maker'] })], (req, res) =>
+router.post('/application-details/:dealId/unissued-facilities/:facilityId/change', [validateToken, validateBank, validateRole({ role: [MAKER] })], (req, res) =>
   postChangeUnissuedFacilityPreview(req, res));
 
+// TODO: DTFS2-6627 - Check whether this needs validation
```
Comment (lmarrai-sw) [inline gef-ui/server/routes/unissued-facilities.js:34]: Flagging TODO

File: gef-ui/server/routes/application-details.js [hunk 7]
```diff
@@ -1,10 +1,11 @@
...
 const { applicationDetails, postApplicationDetails } = require('../controllers/application-details');
-const { validateToken, validateBank } = require('../middleware');
+const { validateToken, validateBank, validateRole } = require('../middleware');
+const { MAKER, CHECKER, READ_ONLY } = require('../constants/roles');
 
 const router = express.Router();
 
-router.get('/application-details/:dealId', [validateToken, validateBank], (req, res) => applicationDetails(req, res));
-router.post('/application-details/:dealId', [validateToken, validateBank], (req, res) => postApplicationDetails(req, res));
+router.get('/application-details/:dealId', [validateToken, validateRole({ role: [MAKER, CHECKER, READ_ONLY] }), validateBank], (req, res) => applicationDetails(req, res));
```
Comment (lmarrai-sw) [inline gef-ui/server/routes/application-details.js:8]: What about `ADMIN`?

File: gef-ui/server/routes/application-details.js [hunk 8]
```diff
@@ -1,10 +1,11 @@
...
+router.post('/application-details/:dealId', [validateToken, validateRole({ role: [MAKER, CHECKER] }), validateBank], (req, res) => postApplicationDetails(req, res));
```
Comment (lmarrai-sw) [inline gef-ui/server/routes/application-details.js:9]: Why does `CHECKER` have permission here?

File: gef-ui/server/middleware/validateRole/index.js [hunk 9]
```diff
@@ -1,23 +1,24 @@
...
+    return false;
+  }
+
+  const userHasOneOfTheRequiredRoles = requiredRoles.some((role) => user.roles.includes(role));
   return userHasOneOfTheRequiredRoles;
 };
 /**
  *
- * @param {object} opts  (i.e. { role: ['maker'] })
+ * @param {object} opts  (i.e. { role: [MAKER] })
```
Comment (lmarrai-sw) [inline gef-ui/server/middleware/validateRole/index.js:21]:
```suggestion
 * @param {object} opts  (i.e. { role: ['maker'] })
```

File: gef-ui/server/middleware/validateRole/index.js [hunk 10]
```diff
@@ -1,23 +1,24 @@
-const ukefAdminRoles = ['ukef_operations', 'admin', 'ukef_admin'];
-
 /**
  *
- * @param {array} requiredRoles  (i.e. ['maker'])
- * @param {array} user (i.e. ['checker'] or ['maker', 'checker'])
+ * @param {array} requiredRoles  (i.e. [MAKER])
+ * @param {array} user (i.e. [CHECKER] or [MAKER, CHECKER])
```
Comment (lmarrai-sw) [inline gef-ui/server/middleware/validateRole/index.js:4]:
```suggestion
 * @param {array} requiredRoles  (i.e. ['maker'])
 * @param {array} user (i.e. ['checker'] or ['maker', 'checker'])
```

File: gef-ui/package.json [hunk 11]
```diff
@@ -19,13 +19,14 @@
   "author": "UK Export Finance",
   "main": "index.js",
   "scripts": {
-    "api-test": "jest --coverage --verbose --forceExit --config=api-test.jest.config.js",
+    "api-test": "jest --verbose --forceExit --config=api-test.jest.config.js",
```
Comment (lmarrai-sw) [inline gef-ui/package.json:22]:
```suggestion
    "api-test": "jest --coverage --verbose --forceExit --config=api-test.jest.config.js",
```
I think we committed this by accident

File: gef-ui/api-tests/common-tests/role-validation-api-tests.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+    if (nonWhitelistedRoles.length) {
+      describe('non-whitelisted roles', () => {
+        it.each(nonWhitelistedRoles)("returns a 302 response if the user only has the '%s' role", async (disallowedRole) => {
+          // const validateRoleSpy = jest.spyOn(middleware, 'validateRole');
+
+          const { sessionCookie } = storage.saveUserSession([disallowedRole]);
+
+          const response = await makeRequestWithHeaders({ Cookie: [`dtfs-session=${encodeURIComponent(sessionCookie)}`] });
+          // expect(validateRoleSpy).toHaveBeenCalledWith({ role: whitelistedRoles });
```
Comment (lmarrai-sw) [inline gef-ui/api-tests/common-tests/role-validation-api-tests.js:65]: watch out for the commented out code here

File: gef-ui/api-tests/common-tests/role-validation-api-tests.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+const storage = require('../test-helpers/storage/storage');
+
+const allRoles = Object.values(ROLES);
+
+const withRoleValidationApiTests = ({
+  makeRequestWithHeaders,
+  whitelistedRoles,
+  successCode,
+  successHeaders,
+  disableHappyPath, // TODO DTFS2-6654: remove and test happy paths.
```
Comment (lmarrai-sw) [inline gef-ui/api-tests/common-tests/role-validation-api-tests.js:22]: Is this the right ticket number?

File: gef-ui/api-tests/about-exporter.api-test.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+const dealId = '123';
+
+describe('about exporter routes', () => {
+  describe('GET /application-details/:dealId/about-exporter', () => {
+    withRoleValidationApiTests({
+      makeRequestWithHeaders: (headers) => get(`/application-details/${dealId}/about-exporter`, {}, headers),
+      whitelistedRoles: [MAKER],
+      successCode: 200,
+      disableHappyPath: true, // TODO DTFS2-6627: remove and test happy path.
```
Comment (lmarrai-sw) [inline gef-ui/api-tests/about-exporter.api-test.js:14]: I think we want a new ticket number for these TODOs, with a new ticket raised explaining the work required. @oscar-richardson-softwire did something similar for portal

File: gef-ui/server/routes/application-details.js (hunk 7 above)
Comment (AlexBramhill) [reply gef-ui/server/routes/application-details.js:8]: @abhi-markan requested 
`application-details/:dealId - Maker, checker, read-only please.`

File: gef-ui/server/routes/application-details.js (hunk 8 above)
Comment (AlexBramhill) [reply gef-ui/server/routes/application-details.js:9]: As above

File: gef-ui/api-tests/about-exporter.api-test.js (hunk 14 above)
Comment (AlexBramhill) [reply gef-ui/api-tests/about-exporter.api-test.js]: Updated

File: portal/package.json (hunk 4 above)
Comment (AlexBramhill) [reply portal/package.json:30]: updated

File: gef-ui/server/routes/unissued-facilities.js (hunk 5 above)
Comment (AlexBramhill) [reply gef-ui/server/routes/unissued-facilities.js]: ty

File: gef-ui/server/routes/application-details.js (hunk 7 above)
Comment (lmarrai-sw) [reply gef-ui/server/routes/application-details.js:8]: Okay, I think we should double-check as it was my understanding that `admin` has the same rights as `read-only` plus being able to edit users

File: gef-ui/server/routes/application-details.js (hunk 8 above)
Comment (lmarrai-sw) [reply gef-ui/server/routes/application-details.js:9]: I'm not convinced by this either - `postApplicationDetails` redirects to `/gef/application-details/${dealId}/submit` and only makers have access to `GET /gef/application-details/${dealId}/submit`

**Review by lmarrai-sw — COMMENTED** [review]
No new comments from me, just the threads from the previous review
Also, you've got a failing UI test (looks like a quick fix)

File: gef-ui/server/routes/application-details.js (hunk 7 above)
Comment (AlexBramhill) [reply gef-ui/server/routes/application-details.js]: updated after checking

File: gef-ui/server/routes/application-details.js (hunk 8 above)
Comment (AlexBramhill) [reply gef-ui/server/routes/application-details.js]: updated

**Review by lmarrai-sw — APPROVED** [review]

**Pull request merged**